### PR TITLE
[RFC] Notify attached UIs whenever menus change

### PR DESCRIPTION
**Pull request by rogual**
This adds a redraw notification "update_menu" which is sent whenever
Vim's menus are changed by the :menu command and friends.

### CODE REVIEW

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -374,6 +375,10 @@ static void tui_cursor_goto(UI *ui, int row, int col)
   unibi_goto(ui, row, col);
 }
 
+static void tui_update_menu(UI *ui)
+{
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: Maybe a comment, like `// Do nothing; menus are for GUI-only`. (Assuming they are GUI-only. Are they?)

**Comment by justinmk** [comment]
Needs a test in `test/functional/api/`.

File: src/nvim/tui/tui.c (hunk 1 above)
Comment (fmoralesc) [inline src/nvim/tui/tui.c]: >  Are they?

Currently, yes.

**Comment by rogual** [comment]
Comment & test added.

**Comment by rogual** [comment]
Changed indentation to 2 spaces.

**Comment by mhinz** [comment]
LGTM

**Comment by rogual** [comment]
Fixed typo in test.

**Comment by rogual** [comment]
Anything else that needs to happen before this can be merged?

**Comment by justinmk** [comment]
Out of the current selection of top-level UI messages, "update menu" seems a bit specific and potentially not future-compatible. Currently the update_menu message sends nothing except a "true" signal. Will we will able to extend it in the future without breaking clients?

Since `args` is a free-form dictionary I suppose the answer is yes, but would like to hear opinions...

**Comment by rogual** [comment]
I did consider providing more information in the notification, but this just seemed to create a necessity for a more complex system of messages detailing exactly what happened; menu_added, menu_removed, menu_text_edited, menu_changed, etc., and a protocol for serializing the details into the notification arguments.

Leaving the client to retrieve menu info itself if it's interested, using for example the bare `:menu` command, allowed me to sidestep the issue.

Though I realize I'm coming at this from the point of view of what's convenient for my own GUI app here :)

**Comment by justinmk** [comment]
I believe a generalized solution here would be to post a message for all `:command` invocations. This of course should be asynchronous. E.g. in this case, the message would be something like:

```
{ "vim_command" : "menu",
  "args" : {
    "command_args" : [ "Test.Test" ]
    "interactive"  : true // Indicates whether the command was
  }                       // invoked by the user or by a script.
}
```

However, this is something that all clients (including non-UI) should be able to listen to. Currently we don't really have a mechanism for non-UI clients to receive non-autocmd messages. @tarruda is there any design reason for this?

**Comment by tarruda** [comment]
> Currently we don't really have a mechanism for non-UI clients to receive non-autocmd messages. @tarruda is there any design reason for this?

The `vim_subscribe`/`vim_subscribe` API functions can be used to listen for msgpack-rpc notifications broadcasts, which is achieved by calling `rpcnotify` with a channel id of `0`.

Sorry for dropping into this issue so late, but this goes against my idea of separating UI and core. What I want is for the UI protocol to be a simple way of notifying clients about screen updates, and not have it extended every time a UI implementation needs to accommodate a new feature. Even though advanced customizations are not possible yet(for compatibility, maybe we should allow msgpack-rpc clients to define lower-case ex commands/functions?), unnecessary extensions to the UI protocol should be avoided. The `:menu` ex command should probably be deleted.

Things like `:menu` should be "injected" by the UI, and not be defined as core commands. For example, if a GUI wants to expose menu customizations to users, one option is to define a `:Menu` ex command via msgpack-rpc, something that is very well possible since remote UIs use the same infrastructure as remote plugins.

Another example: If your GUI wants to expose font customization to the user, it can implement an ex command such as `:SetFont` that is autocompleted using a vimscript function also implemented by the UI. Such things are the goal of the remote plugin infrastructure and you can see some examples in the python host/client, but the functions to define remote commands/functions/autocmds are available to any msgpack-rpc client.

**Comment by justinmk** [comment]
`:menu` is basically a mapping. Can you expand on how it adds coupling to the UI?

**Comment by tarruda** [comment]
> :menu is basically a mapping. Can you expand on how it adds coupling to the UI?

Since I never have used `:menu`, I had the impression it was a GUI-specific feature, but after reading the help I can see you are right. Please disregard my previous comment.

**Comment by rogual** [comment]
Now that @tarruda's objection has been addressed, would it be possible to get this merged, while we wait for a more general solution?

Right now AFAICT this is the only way of adding menu support to clients.

**Comment by tarruda** [comment]
@rogual do a rebase and I will merge

**Comment by rogual** [comment]
Rebased.

**Comment by tarruda** [comment]
@rogual it is not rebased yet, the parent commit on this branch is https://github.com/rogual/neovim/commit/ccebc1f8d3b30536bbaab532cc5acc7ba7ec3d4b while master is https://github.com/neovim/neovim/commit/250aca4f8938ded5bf7dfbc56904b4e696d219b1

**Comment by rogual** [comment]
Oops, sorry! Rebased properly now.

**Comment by tarruda** [comment]
Thanks

**Comment by rogual** [comment]
Thank you!

**Pull request merged**